File: packages/protocols/x402-ampersend-sdk/python/ampersend-sdk/src/ampersend_sdk/x402/wallets/smart_account/exact.py
"""
Smart account signing utilities for ERC-1271 signatures.

This module provides functions to sign typed data and ERC-3009 authorizations
using smart accounts (Safe + OwnableValidator) instead of plain EOA accounts.

The signatures are compatible with ERC-1271 validation and can be verified
on-chain by smart contract wallets.
"""

from x402.chains import get_chain_id
from x402.common import x402_VERSION
from x402.exact import prepare_payment_header
from x402_a2a import PaymentPayload, PaymentRequirements
from x402_a2a.types import (
    EIP3009Authorization,
    ExactPaymentPayload,
)

from ....smart_account.sign import SmartAccountConfig, smart_account_sign_typed_data


def sign_erc3009_authorization(
    config: SmartAccountConfig,
    authorization: EIP3009Authorization,
    domain_verifying_contract: str,
    domain_chain_id: int,
    domain_name: str,
    domain_version: str,
) -> str:
    """
    Sign an ERC-3009 transferWithAuthorization for a smart account.

    ERC-3009 is used by USDC and other tokens to allow gasless transfers
    via signed authorizations. This function creates a signature that
    can be used with the token's transferWithAuthorization function.

    Args:
        session_key: Agent's private key
        smart_account_address: Smart account address
        authorization: Authorization data
        domain_verifying_contract: Verifying contract address
        domain_chain_id: Chain ID (e.g., 84532 for Base Sepolia)
        validator_address: OwnableValidator address (default: deployed address)

    Returns:
        ERC-1271 encoded signature
    """
    # ERC-3009 domain
    domain = {
        "name": domain_name,
        "version": domain_version,
        "chainId": domain_chain_id,
        "verifyingContract": domain_verifying_contract,
    }

    types = {
        "EIP712Domain": [
            {"name": "name", "type": "string"},
            {"name": "version", "type": "string"},
            {"name": "chainId", "type": "uint256"},
            {"name": "verifyingContract", "type": "address"},
        ],
        "TransferWithAuthorization": [
            {"name": "from", "type": "address"},
            {"name": "to", "type": "address"},
            {"name": "value", "type": "uint256"},
            {"name": "validAfter", "type": "uint256"},
            {"name": "validBefore", "type": "uint256"},
            {"name": "nonce", "type": "bytes32"},
        ],
    }

    return smart_account_sign_typed_data(
        config=config,
        domain=domain,
        types=types,
        message=authorization.model_dump(by_alias=True),
        primary_type="TransferWithAuthorization",
    )


def smart_account_create_payment(
    requirements: PaymentRequirements,
    config: SmartAccountConfig,
) -> PaymentPayload:
    """
    Create a payment payload with smart account signature.

    This function constructs a complete x402 PaymentPayload for the exact scheme
    using a smart account signature. It handles nonce generation, timestamp
    calculation, authorization data creation, signing, and payload construction.

    Used internally by smart_account_create_payment factory and can be called
    directly for advanced use cases.

    Args:
        requirements: Single PaymentRequirements from x402 server
        smart_account_address: Smart account address
        agent_private_key: Agent key private key (owner of smart account)
        chain_id: Chain ID (e.g., 84532 for Base Sepolia)
        validator_address: OwnableValidator address (default: deployed address)

    Returns:
        PaymentPayload ready to submit to x402 service

    Raises:
        ValueError: If unsupported payment scheme
    """

    if requirements.scheme != "exact":
        raise ValueError(f"Unsupported payment scheme: {requirements.scheme}")

    # TODO: clean this up once x402 lib improves
    # Handle nonce conversion for x402.exact compatibility
    unsigned_payload = prepare_payment_header(
        sender_address=config.smart_account_address,
        x402_version=x402_VERSION,
        payment_requirements=requirements,
    )
    nonce_raw = unsigned_payload["payload"]["authorization"]["nonce"]
    if isinstance(nonce_raw, bytes):
        unsigned_payload["payload"]["authorization"]["nonce"] = "0x" + nonce_raw.hex()

    # Construct authorization
    authorization = EIP3009Authorization.model_validate(
        unsigned_payload["payload"]["authorization"], by_alias=True
    )

    assert requirements.extra is not None
    signature = sign_erc3009_authorization(
        config=config,
        authorization=authorization,
        domain_verifying_contract=requirements.asset,
        domain_chain_id=int(get_chain_id(requirements.network)),
        domain_name=requirements.extra["name"],
        domain_version=requirements.extra["version"],
    )

    exact_payload = ExactPaymentPayload(
        signature=signature,
        authorization=authorization,
    )

    return PaymentPayload(
        x402_version=1,
        scheme="exact",
        network=requirements.network,
        payload=exact_payload,
    )

File: packages/protocols/x402-ampersend-sdk/typescript/packages/ampersend-sdk/src/x402/wallets/smart-account/exact.ts
import { toHex, type Address, type Hex } from "viem"
import type { PaymentPayload, PaymentRequirements } from "x402/types"

import { signERC3009Authorization } from "../../../smart-account/index.ts"

/**
 * Generates a random 32-byte nonce for use in authorization signatures
 */
function createNonce(): Hex {
  const cryptoObj =
    typeof globalThis.crypto !== "undefined" && typeof globalThis.crypto.getRandomValues === "function"
      ? globalThis.crypto
      : // Dynamic require is needed to support node.js
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        require("crypto").webcrypto
  return toHex(cryptoObj.getRandomValues(new Uint8Array(32)))
}

/**
 * Configuration for creating an exact payment with ERC-3009
 */
export interface ExactPaymentConfig {
  /** Smart account address (payment sender) */
  smartAccountAddress: Address
  /** Session key private key for signing */
  sessionKeyPrivateKey: Hex
  /** Chain ID for the blockchain network */
  chainId: number
  /** OwnableValidator address for ERC-1271 validation */
  validatorAddress: Address
}

/**
 * Creates a payment payload using the "exact" scheme with ERC-3009 USDC authorization
 *
 * This implements the x402 "exact" payment scheme, which uses USDC's transferWithAuthorization
 * (ERC-3009) to create signed payment authorizations. The signature is created using ERC-1271
 * from a smart account via the OwnableValidator module.
 *
 * @param requirements - Payment requirements from the x402 server
 * @param config - Configuration for the smart account wallet
 * @returns Payment payload ready to send to x402 server
 * @throws Error if payment requirements are invalid or signing fails
 */
export async function createExactPayment(
  requirements: PaymentRequirements,
  config: ExactPaymentConfig,
): Promise<PaymentPayload> {
  // Generate nonce and validity timestamps
  const nonce = createNonce()
  const validAfter = BigInt(Math.floor(Date.now() / 1000) - 600) // 10 minutes before
  const validBefore = BigInt(Math.floor(Date.now() / 1000) + requirements.maxTimeoutSeconds)

  // Prepare authorization data for ERC-3009 signing
  const authData = {
    from: config.smartAccountAddress,
    to: requirements.payTo as Address,
    value: BigInt(requirements.maxAmountRequired),
    validAfter,
    validBefore,
    nonce,
  }

  // Get domain params from requirements.extra (provided by server)
  const domainName = requirements.extra?.name as string | undefined
  const domainVersion = requirements.extra?.version as string | undefined

  if (!domainName || !domainVersion) {
    throw new Error("requirements.extra must contain 'name' and 'version' for EIP-712 domain")
  }

  // Sign using ERC-1271 with OwnableValidator
  const signature = await signERC3009Authorization(
    config.sessionKeyPrivateKey,
    config.smartAccountAddress,
    authData,
    requirements.asset as Address,
    config.chainId,
    config.validatorAddress,
    domainName,
    domainVersion,
  )

  // Construct payment payload matching x402 exact scheme format
  const paymentPayload: PaymentPayload = {
    x402Version: 1,
    scheme: "exact" as const,
    network: requirements.network,
    payload: {
      signature: signature as string,
      authorization: {
        from: config.smartAccountAddress as string,
        to: requirements.payTo,
        value: requirements.maxAmountRequired,
        validAfter: validAfter.toString(),
        validBefore: validBefore.toString(),
        nonce: nonce as string,
      },
    },
  }

  return paymentPayload
}
